### PR TITLE
Fix #384

### LINKED ISSUE #384: PowerShellToolsProcessHost visible when launching Visual Studio
When I launch Visual Studio a small console window, which I'm guessing is `PowerShellToolsProcessHost`, flashes in front of VS and steals focus from the VS gui.

Please hide this window if possible...

### PR DESCRIPTION
**Pull request by ericdevg**
#384

Highlights:
1. make powershell host process as wpf application, on top but invisible.
2. attaching an console to wpf application to handle stdin

@AndreSayreMSFT @HuanhuanSunMSFT @Microsoft/poshtools 

NOTE: I realized that github is doing a slightly better job on giving diff on this specific change comparing with code flow, but feel free to pick up whichever you like.

### CODE REVIEW

**Comment by ericdevg** [comment]
Ping ... @AndreSayreMSFT @HuanhuanSunMSFT

File: PowerShellTools/Manifest/14.0/source.extension.vsixmanifest [hunk 1]
```diff
@@ -35,5 +35,6 @@
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostService" Path="|PowerShellTools.HostService|" AssemblyName="|PowerShellTools.HostService;AssemblyName|" />
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostServicex86" Path="|PowerShellTools.HostServicex86|" AssemblyName="|PowerShellTools.HostServicex86;AssemblyName|" />
```
Comment (ASayre) [inline PowerShellTools/Manifest/14.0/source.extension.vsixmanifest]: HostServicex86 -> HostService.x86

File: PowerShellTools/Manifest/14.0/source.extension.vsixmanifest [hunk 2]
```diff
@@ -35,5 +35,6 @@
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostService" Path="|PowerShellTools.HostService|" AssemblyName="|PowerShellTools.HostService;AssemblyName|" />
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostServicex86" Path="|PowerShellTools.HostServicex86|" AssemblyName="|PowerShellTools.HostServicex86;AssemblyName|" />
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" Path="|PowerShellTools.Common|" AssemblyName="|PowerShellTools.Common;AssemblyName|" d:ProjectName="%CurrentProject%.Common" />
+    <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="PowerShell.HostService.Console" Path="|PowerShell.HostService.Console|" AssemblyName="|PowerShell.HostService.Console;AssemblyName|" />
```
Comment (ASayre) [inline PowerShellTools/Manifest/14.0/source.extension.vsixmanifest]: d:ProjectName should be |%currentProject...

File: PowerShellTools/Manifest/11.0/source.extension.vsixmanifest [hunk 3]
```diff
@@ -31,7 +31,8 @@
...
     <Asset Type="Microsoft.VisualStudio.MefComponent" d:Source="Project" d:ProjectName="%CurrentProject%.Repl" Path="|PowerShellTools.Repl|" />
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostService" Path="|PowerShellTools.HostService|" AssemblyName="|PowerShellTools.HostService;AssemblyName|" />
-    <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="%CurrentProject%.HostServicex86" Path="|PowerShellTools.HostServicex86|" AssemblyName="|PowerShellTools.HostServicex86;AssemblyName|" />
+    <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" Path="|PowerShellTools.HostService.x86|" AssemblyName="|PowerShellTools.HostService.x86;AssemblyName|" />
     <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" Path="|PowerShellTools.Common|" AssemblyName="|PowerShellTools.Common;AssemblyName|" d:ProjectName="%CurrentProject%.Common" />
+    <Asset Type="Microsoft.VisualStudio.Assembly" d:Source="Project" d:ProjectName="PowerShell.HostService.Console" Path="|PowerShell.HostService.Console|" AssemblyName="|PowerShell.HostService.Console;AssemblyName|" />
```
Comment (ASayre) [inline PowerShellTools/Manifest/11.0/source.extension.vsixmanifest]: d:projectName

File: PowerShellTools.HostService/ConsoleManager.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Create console process
+        /// </summary>
+        /// <returns>Console process created</returns>
+        private static Process CreateConsole()
+        {
+            Process p = new Process(); ;
```
Comment (ASayre) [inline PowerShellTools.HostService/ConsoleManager.cs]: Nit: extra ;

File: PowerShell.HostService.Console/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("PowerShellTools Host Service Console")]
+[assembly: AssemblyDescription("Console of host Service for the PowerShell Tools in Visual Studio")]
```
Comment (ASayre) [inline PowerShell.HostService.Console/Properties/AssemblyInfo.cs]: Nit: Host Service Console for the PowerShell Tools in Visual Studio

File: PowerShell.HostService.Console/Program.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace PowerShell.HostService.Console
+{
+    class Program
```
Comment (ASayre) [inline PowerShell.HostService.Console/Program.cs]: Nit: Access modifiers

File: PowerShell.HostService.Console/Program.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        private static AutoResetEvent _processExitEvent = new AutoResetEvent(false);
+
+        static void Main(string[] args)
```
Comment (ASayre) [inline PowerShell.HostService.Console/Program.cs]: Nit: Access modifiers

File: PowershellTools.Common/Constants.cs [hunk 8]
```diff
@@ -31,6 +33,9 @@ internal static class Constants
         public const string UniqueEndpointArg = "/endpoint:";
         public const string ReadyEventUniqueNameArg = "/readyEventUniqueName:";
 
+        // Arguments for vspowershellhostconsole.exe
+        public const string ConsoleProcessIdArg = "/hpPid:";
```
Comment (ASayre) [inline PowershellTools.Common/Constants.cs:37]: /hp?  [HostPid?]

File: PowerShell.HostService.Console/Program.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            try
+            {
+                int _powerShellHostProcessId;
```
Comment (ASayre) [inline PowerShell.HostService.Console/Program.cs:27]: Validate argument against ConsoleProcessIdArg?

File: PowerShell.HostService.Console/Program.cs [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                    p.Exited += new EventHandler(
+                        (sender, eventArgs) =>
+                        {
+                            _processExitEvent.Set();
+                        });
+                }
+
+                _processExitEvent.WaitOne();
+            }
+            catch (Exception)
```
Comment (ASayre) [inline PowerShell.HostService.Console/Program.cs]: nit:catch{}

File: PowerShellTools.HostService/App.xaml.cs [hunk 11]
```diff
@@ -100,31 +108,14 @@ internal static int Main(string[] args)
...
                             }
 
-                            _processExitEvent.Set();
+                            Environment.Exit(0);
                         });
                 }
-
-                _processExitEvent.WaitOne();
             }
             catch (Exception)
```
Comment (ASayre) [inline PowerShellTools.HostService/App.xaml.cs]: nit: catch{}

File: PowerShellTools.HostService/App.xaml.cs [hunk 12]
```diff
@@ -1,62 +1,70 @@
...
+            EndpointGuid = e.Args[0].Remove(0, Constants.UniqueEndpointArg.Length);
             if (EndpointGuid.Length != Guid.Empty.ToString().Length)
             {
-                return 1;
+                return;
             }
 
             int vsProcessId;
-            if (!Int32.TryParse(args[1].Remove(0, Constants.VsProcessIdArg.Length),
+            if (!Int32.TryParse(e.Args[1].Remove(0, Constants.VsProcessIdArg.Length),
```
Comment (ASayre) [inline PowerShellTools.HostService/App.xaml.cs]: nit: int.TryParse

File: PowerShell.HostService.Console/Program.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    class Program
+    {
+        private static AutoResetEvent _processExitEvent = new AutoResetEvent(false);
+
+        static void Main(string[] args)
+        {
+            try
+            {
+                int _powerShellHostProcessId;
+                if (!Int32.TryParse(args[0].Remove(0, Constants.ConsoleProcessIdArg.Length),
```
Comment (ASayre) [inline PowerShell.HostService.Console/Program.cs]: nit:int.TryParse

File: PowerShellTools.HostService/ConsoleManager.cs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        /// </summary>
+        public static bool HasConsole
+        {
+            get { return GetConsoleWindow() != IntPtr.Zero; }
+        }
+
+        /// <summary>
+        /// Creates a new console instance if the process is not attached to a console already.
+        /// </summary>
+        public static void AttachConsole()
```
Comment (ASayre) [inline PowerShellTools.HostService/ConsoleManager.cs]: do these need to be public?

File: PowerShellTools.HostService/ConsoleManager.cs [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        private static extern IntPtr GetConsoleWindow();
+
+        [DllImport(Kernel32_DllName)]
+        public static extern bool AttachConsole(uint dwProcessId);
+
+        /// <summary>
+        /// Check if there is already a console attached.
+        /// One application can only have one console attached.
+        /// </summary>
+        public static bool HasConsole
```
Comment (ASayre) [inline PowerShellTools.HostService/ConsoleManager.cs]: Public?

File: PowerShellTools/Manifest/14.0/source.extension.vsixmanifest (hunk 1 above)
Comment (ericdevg) [inline PowerShellTools/Manifest/14.0/source.extension.vsixmanifest]: Fixed.

File: PowerShellTools/Manifest/14.0/source.extension.vsixmanifest (hunk 2 above)
Comment (ericdevg) [inline PowerShellTools/Manifest/14.0/source.extension.vsixmanifest]: Fixed.

File: PowerShellTools/Manifest/11.0/source.extension.vsixmanifest (hunk 3 above)
Comment (ericdevg) [inline PowerShellTools/Manifest/11.0/source.extension.vsixmanifest]: Fixed.

File: PowerShellTools.HostService/ConsoleManager.cs (hunk 4 above)
Comment (ericdevg) [inline PowerShellTools.HostService/ConsoleManager.cs]: Fixed.

File: PowerShell.HostService.Console/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (ericdevg) [inline PowerShell.HostService.Console/Properties/AssemblyInfo.cs]: Fixed.

File: PowershellTools.Common/Constants.cs (hunk 8 above)
Comment (ericdevg) [inline PowershellTools.Common/Constants.cs:37]: Trying to keep it consistent with /vsPid:

File: PowerShell.HostService.Console/Program.cs (hunk 9 above)
Comment (ericdevg) [inline PowerShell.HostService.Console/Program.cs:27]: Fixed.

File: PowerShell.HostService.Console/Program.cs (hunk 10 above)
Comment (ericdevg) [inline PowerShell.HostService.Console/Program.cs]: Fixed.

File: PowerShellTools.HostService/App.xaml.cs (hunk 11 above)
Comment (ericdevg) [inline PowerShellTools.HostService/App.xaml.cs]: Fixed.

File: PowerShellTools.HostService/ConsoleManager.cs (hunk 14 above)
Comment (ericdevg) [inline PowerShellTools.HostService/ConsoleManager.cs]: Fixed.

File: PowerShellTools.HostService/ConsoleManager.cs (hunk 15 above)
Comment (ericdevg) [inline PowerShellTools.HostService/ConsoleManager.cs]: Fixed.

**Comment by ASayre** [comment]
Looks good!

**Pull request merged**